aws/resource_aws_iam_role_test.go

```diff
@@ -491,321 +494,1411 @@ func TestAccAWSIAMRole_tags(t *testing.T) {
+		}
+
+		_, err = conn.DetachRolePolicy(&iam.DetachRolePolicyInput{
+			PolicyArn: aws.String(managedARN),
+			RoleName:  role.Role.RoleName,
+		})
+
+		if err != nil {
+			return err
+		}
+
+		return nil
+	}
+}
+
+func testAccCheckAWSRolePolicyAttachManagedPolicy(role *iam.GetRoleOutput, policyName string) resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+		conn := testAccProvider.Meta().(*AWSClient).iamconn
+
+		var managedARN string
+		input := &iam.ListPoliciesInput{
+			PathPrefix:        aws.String("/tf-testing/"),
+			PolicyUsageFilter: aws.String("PermissionsPolicy"),
+			Scope:             aws.String("Local"),
+		}
+
+		err := conn.ListPoliciesPages(input, func(page *iam.ListPoliciesOutput, lastPage bool) bool {
+			for _, v := range page.Policies {
+				if *v.PolicyName == policyName {
+					managedARN = *v.Arn
+					break
+				}
+			}
+			return !lastPage
+		})
+		if err != nil && !tfawserr.ErrCodeEquals(err, iam.ErrCodeNoSuchEntityException) {
+			return fmt.Errorf("finding managed policy (%s): %w", policyName, err)
+		}
+		if managedARN == "" {
+			return fmt.Errorf("managed policy (%s) not found", policyName)
+		}
+
+		_, err = conn.AttachRolePolicy(&iam.AttachRolePolicyInput{
+			PolicyArn: aws.String(managedARN),
+			RoleName:  role.Role.RoleName,
+		})
+		if err != nil {
+			return err
+		}
+
+		return nil
+	}
+}
+
+func testAccCheckAWSRolePolicyAddInlinePolicy(role *iam.GetRoleOutput, inlinePolicy string) resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+		conn := testAccProvider.Meta().(*AWSClient).iamconn
+
+		_, err := conn.PutRolePolicy(&iam.PutRolePolicyInput{
+			PolicyDocument: aws.String(testAccAWSRolePolicyExtraInlineConfig()),
+			PolicyName:     aws.String(inlinePolicy),
+			RoleName:       aws.String(*role.Role.RoleName),
+		})
+
+		if err != nil {
+			return err
+		}
+		return nil
+	}
+}
+
+func testAccCheckAWSRolePolicyRemoveInlinePolicy(role *iam.GetRoleOutput, inlinePolicy string) resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+		conn := testAccProvider.Meta().(*AWSClient).iamconn
+
+		_, err := conn.DeleteRolePolicy(&iam.DeleteRolePolicyInput{
+			PolicyName: aws.String(inlinePolicy),
+			RoleName:   aws.String(*role.Role.RoleName),
+		})
+
+		if err != nil {
+			return err
+		}
+		return nil
+	}
+}
+
+func compareStringSlices(a []string, b []string) bool {
+	if len(a) != len(b) {
+		return false
+	}
+	sort.Strings(a)
+	sort.Strings(b)
+	for i := range b {
+		if a[i] != b[i] {
+			return false
+		}
+	}
+	return true
+}
+
+func testAccCheckIAMRoleConfig_MaxSessionDuration(rName string, maxSessionDuration int) string {
+	return fmt.Sprintf(`
+resource "aws_iam_role" "test" {
+  name                 = "test-role-%s"
+  path                 = "/"
+  max_session_duration = %d
+
+  assume_role_policy = <<EOF
+{
+  "Version": "2012-10-17",
+  "Statement": [
+    {
+      "Effect": "Allow",
+      "Principal": {
+        "Service": [
+          "ec2.amazonaws.com"
```

Comment:
Here and below should be using the `aws_partition` data source:

```suggestion
          "ec2.${data.aws_partition.current.dns_suffix}"
```

Reply:
Fixed all